src/Dfc.CourseDirectory.Core/Models/Postcode.cs

```diff
@@ -0,0 +1,79 @@
+﻿using System;
+using System.Text.RegularExpressions;
+
+namespace Dfc.CourseDirectory.Core.Models
+{
+    public sealed class Postcode : IEquatable<Postcode>
+    {
+        private static readonly Regex _ukPostcodePattern = new Regex(
+            @"^((([A-Za-z][0-9]{1,2})|(([A-Za-z][A-Ha-hJ-Yj-y][0-9]{1,2})|(([A-Za-z][0-9][A-Za-z])|([A-Za-z][A-Ha-hJ-Yj-y][0-9][A-Za-z]?))))\s?[0-9][A-Za-z]{2})$",
+            RegexOptions.Compiled);
+
+        private readonly string _value;
+
+        public Postcode(string value)
+            : this(value, validated: false)
+        {
+        }
+
+        private Postcode(string value, bool validated)
+        {
+            if (!validated)
+            {
+                if (!ValidateAndNormalize(value, out value))
+                {
+                    throw new ArgumentException("Input is not a valid UK postcode.", nameof(value));
+                }
+            }
+
+            _value = value ?? throw new ArgumentNullException(nameof(value));
+        }
+
+        public static bool TryParse(string input, out Postcode postcode)
+        {
+            if (!ValidateAndNormalize(input, out var normalized))
+            {
+                postcode = default;
+                return false;
+            }
+
+            postcode = new Postcode(normalized, validated: true);
+            return true;
+        }
+
+        private static bool ValidateAndNormalize(string input, out string normalized)
+        {
+            if (input == null)
```

Comment:
Check `string.IsNullOrWhiteSpace(input)`?

Reply:
The Regex check below handles all that. This is only here to avoid the `ArgumentNullException` on `Regex.IsMatch()`

Reply:
Missing an opportunity for an early out without having to invoke an expensive regex?